src/historywork/DownloadApplyTxsWork.cpp

```diff
@@ -0,0 +1,123 @@
+#include "ledger/LedgerManager.h"
+#include "work/WorkSequence.h"
+
+namespace stellar
+{
+
+DownloadApplyTxsWork::DownloadApplyTxsWork(
+    Application& app, TmpDir const& downloadDir, LedgerRange range,
+    LedgerHeaderHistoryEntry& lastApplied)
+    : BatchWork(app, "download-apply-ledgers")
+    , mRange(range)
+    , mDownloadDir(downloadDir)
+    , mLastApplied(lastApplied)
+    , mCheckpointToQueue(
+          app.getHistoryManager().checkpointContainingLedger(range.mFirst))
+{
+}
+
+std::shared_ptr<BasicWork>
+DownloadApplyTxsWork::yieldMoreWork()
+{
+    if (!hasNext())
+    {
+        CLOG(ERROR, "Work")
+            << getName() << " has no more children to iterate over! ";
+        return nullptr;
+    }
+
+    CLOG(INFO, "History") << "Downloading, unzipping and applying "
+                          << HISTORY_FILE_TYPE_TRANSACTIONS
+                          << " for checkpoint " << mCheckpointToQueue;
+    FileTransferInfo ft(mDownloadDir, HISTORY_FILE_TYPE_TRANSACTIONS,
+                        mCheckpointToQueue);
+    auto getAndUnzip = std::make_shared<GetAndUnzipRemoteFileWork>(mApp, ft);
+    auto apply = std::make_shared<ApplyCheckpointWork>(
+        mApp, mDownloadDir, mCheckpointToQueue, mRange.mLast);
+
+    std::vector<std::shared_ptr<BasicWork>> seq{getAndUnzip, apply};
+    std::vector<ConditionFn> cond;
+
+    if (!mRunning.empty())
+    {
+        auto prevWork = mRunning.back();
+        auto predicate = [prevWork]() {
+            if (!prevWork)
+            {
+                throw std::runtime_error("Download and apply txs: related Work "
+                                         "is destroyed unexpectedly");
+            }
+            return prevWork->getState() == State::WORK_SUCCESS;
+        };
+        cond = {nullptr, predicate};
+    }
+
+    auto nextWork = std::make_shared<WorkSequence>(
+        mApp, "download-apply-" + std::to_string(mCheckpointToQueue), seq, cond,
+        RETRY_NEVER);
+    mRunning.push_back(nextWork);
+    mCheckpointToQueue += mApp.getHistoryManager().getCheckpointFrequency();
+
+    return nextWork;
+}
+
+void
+DownloadApplyTxsWork::wakeWaitingWork()
+{
+    while (!mRunning.empty())
+    {
+        auto work = mRunning.front();
+        if (work->getState() == State::WORK_SUCCESS)
+        {
+            mRunning.pop_front();
+            continue;
+        }
+
+        if (work->getState() != State::WORK_FAILURE)
+        {
+            // At this point, sequence download->apply is either:
+            // - Still downloading files, in which case `wakeUp` will have
+            // no effect, and it will simply proceed to applying when done
+            // since the condition has been met OR
+            // - Finished downloading, but is waiting for previous checkpoint
+            // to finish applying; in this case, wakeUp will put it back into
+            // RUNNING state and the sequence can proceed with applying
+            work->wakeUp();
```

Comment:
Calling `wakeUp` directly seems to lead to unspecified behavior: for example, if you call `wakeUp` on `RunCommand`, it will respawn a process from what I see (as it expects to only be awaken when the process completes).

Reason is that I think it's actually wrong to call `wakeUp` on arbitrary `BasicWork`: the only mechanism we have for callbacks is to pass a callback to `startWork`, callback that gets invoked when the work wakes up as defined by itself.

I think that if we want to be able to wake up other work, we need to create a new type of work where the contract allows it to be awoken like that.

I think the closest to this would be what I mentioned before, `ConditionalWork` (that waits on some condition to become true before triggering some other work), we can then safely call `doWork` even if the condition is not met.
If we want to avoid the "busy wait" (that sleeps between checks against the condition), we have many options. For example, we can register all instances of `conditionalWork` with some object (like the scheduler), and wake up all conditional work when "something interesting" happens (could be a combination of timeout and explicit calls to `wakeupConditionalWork` when an instance of `DownloadApplyTxsWork` completes for example).
This would also avoid calling the `wake up all` from inside `DownloadApplyTxWork` that is a bit hacky.